File: test/fixtures/external_proof/rust_bitcoin_weightceil_scalar.c
/* Scalar C11 extract of rust-bitcoin Amount::from_sat and a small-product
 * integer ceil (CC0-1.0 semantics). No pointers, no structs: eligible for
 * src/lin_from_c.lin v2. Weight ceil of sats times 4e6 needs 128-bit product and
 * lives in rust_bitcoin_weightceil_c11.c, not this subset.
 * Logical and in a signature must not be classified as a C++ reference.
 */
int64_t rwc_max_money(void) {
    return 21000000 * 100000000;
}

int64_t rwc_mvb_per_wu(void) {
    return 4000000;
}

int64_t rwc_from_sat(int64_t satoshi) {
    if (satoshi < 0) {
        return -1;
    }
    if (satoshi > 21000000 * 100000000) {
        return -1;
    }
    return satoshi;
}

int64_t rwc_div_ceil_small(int64_t prod, int64_t wu) {
    int64_t q;
    int64_t r;
    int64_t n;
    if (prod < 0) {
        return -1;
    }
    if (wu <= 0) {
        return -1;
    }
    q = prod / wu;
    r = prod - q * wu;
    if (r == 0) {
        return q;
    }
    n = q + 1;
    if (n <= 0) {
        return -1;
    }
    return n;
}

File: test/fixtures/external_proof/rust_bitcoin_weightceil_scalar.rs
// Scalar no_std extract for lin_from_rust v2. CC0-1.0 semantics from
// rust-bitcoin units Amount::div_by_weight_ceil
// @ 1cbf4bd62ea99c558c6d8ef794edbd984a2a4684.
// Underscores, &&, and || must survive (v1 treated && as a borrow).
// Small-product ceil only: full (sats*4_000_000) needs 128-bit and lives
// in lin_rust_bitcoin_weightceil.lin, not this subset.

pub fn rwc_max_money() -> u64 {
    return 21_000_000 * 100_000_000;
}

pub fn rwc_mvb_per_wu() -> u64 {
    return 4_000_000;
}

pub fn rwc_from_sat(satoshi: u64) -> i64 {
    if satoshi > 21_000_000 * 100_000_000 {
        return -1;
    }
    return satoshi;
}

pub fn rwc_div_ceil_small(prod: u64, wu: u64) -> i64 {
    if wu == 0 || (prod > 21_000_000 * 100_000_000 && wu < 2) {
        return -1;
    }
    if wu == 0 {
        return -1;
    }
    let q = prod / wu;
    let r = prod - q * wu;
    if r == 0 {
        return q;
    }
    return q + 1;
}
